### ISSUE TITLE
Warning: No tests found in *.dll

### ISSUE THREAD

**Issue by agray** [nunit/nunit#519]
I have a simple test assembly dependent on NUnit 3 Beta 1 nuget package.

I have the "NUnit3 Beta 1"  Framework package installed on my computer.

My Test class is:

    [TestFixture]
    public class GuineaPigTests : TestBase {
        [MethodImpl(MethodImplOptions.Synchronized)]
        [Test]
        public void PageTitle() {
            Driver.Navigate().GoToUrl("https://saucelabs.com/test/guinea-pig");

            // verify the page title is correct
            Assert.IsTrue(Driver.Title.Contains("I am a page title - Sauce Labs"));
        }
    }

Jenkins 1.599 is executing the tests with the following Windows batch command:

    "C:\Program Files (x86)\NUnit.org\bin\nunit-console.exe" NUnit\Version3\MyTestApp\bin\Release\MyTestApp.dll 
    exit %%ERRORLEVEL%%
    
Jenkins console output is:

    NUnit Console Runner 3.0.5509
    Copyright (C) 2014 Charlie Poole

    Test Files:
       NUnit\Version3\MyTestApp\bin\Release\MyTestApp.dll

    Runtime Environment
       OS Version: Microsoft Windows NT 6.3.9600.0
       CLR Version: 4.0.30319.34209

    Options
        ProcessModel: Default    DomainUsage: Default
        Execution Runtime: Not Specified
        Work Directory: C:\Jenkins\workspace\MyTestApp
        Internal Trace: Off

    Test Run Summary
       Overall result: Passed
       Tests run: 0, Passed: 0, Errors: 0, Failures: 0, Inconclusive: 0
       Not run: 0, Invalid: 0, Ignored: 0, Skipped: 0
       Start time: 2015-02-28 08:42:21Z
       End time: 2015-02-28 08:42:21Z
       Duration: 0.083 seconds

    Warning: No tests found in MyTestApp.dll

No exception is being thrown.

Why weren't my tests found?

**Comment by CharliePoole** [comment]
I can't tell from the excepted code but the most common reason is that the
class itself isn't public.

If that's not it, we'll need more info to treat this as an issue.

**Comment by agray** [comment]
Please see my updated description.  Let me know if you require any more details.

What else can I provide to get to the bottom of this issue.

**Comment by CharliePoole** [comment]
Let's leave Jenkins aside for a moment. What happens if you run it directly?

**Comment by agray** [comment]
If I run the same command from a console I get the same result, same output

**Comment by CharliePoole** [comment]
No output header? I'd like to verify what version of nunit-console is executing.

**Comment by agray** [comment]
Have update the console output section of the issue description above.

**Comment by agray** [comment]
Any update on this issue?

**Comment by CharliePoole** [comment]
Sorry, this seems to have slipped through the cracks. What's in your base class TestBase?

**Comment by agray** [comment]
It is a class I wrote (https://www.nuget.org/packages/Saucery2/).

Saucery handles the DesiredCapabilities.  Details here: fullcirclesolutions.com.au.

Working perfectly for NUnit 2.6.4.  Issues are with NUnit 3.

**Comment by agray** [comment]
To avoid this slipping through the cracks again could this please be added to an upcoming release milestone.

**Comment by CharliePoole** [comment]
That's not how we use milestones. It needs to be categorized as a bug and assigned to someone to get worked on. Right now we're trying to confirm the problem.

It seems likely that something is going wrong in the OneTimeSetUp, causing the test to not be run. This might be reported more clearly by the beta release. Have you run it under the beta?

Can you confirm that your test code is ever reached? Can you duplicate the code problem without inheriting from TestBase?

Since the base class is part of a commercial product, we can't debug through it or even reproduce the error.

**Comment by agray** [comment]
The commercial product is mine. I'll add some debug statements and report back here.

**Comment by agray** [comment]
Still not working.

I have NUnit.Framework.3.0.0-beta-1.msi installed on my machine (Window 8.1 Enterprise) from https://github.com/nunit/nunit/releases.

I am using the following command to execute my test assembly.

"C:\Program Files (x86)\NUnit.org\bin\nunit-console.exe" MyTestAssembly.dll

Output is as follows:
NUnit Console Runner 3.0.5561 
Copyright (C) 2014 Charlie Poole

Test Files
    NUnit\Version3\Saucery3Tester\bin\Release\Saucery3Tester.dll

Runtime Environment
   OS Version: Microsoft Windows NT 6.3.9600.0
  CLR Version: 4.0.30319.34209

Options
    ProcessModel: Default    DomainUsage: Default
    Execution Runtime: Not Specified
    Work Directory: C:\Jenkins\workspace\Saucery3NUnit
    Internal Trace: Off

Test Run Summary
    Overall result: Passed
   Tests run: 0, Passed: 0, Errors: 0, Failures: 0, Inconclusive: 0
     Not run: 0, Invalid: 0, Ignored: 0, Skipped: 0
  Start time: 2015-04-09 11:17:31Z
    End time: 2015-04-09 11:17:31Z
    Duration: 0.088 seconds

Results (nunit3) saved as TestResult.xml

**Comment by CharliePoole** [comment]
What does TestResult.xml say, if anything?

**Comment by CharliePoole** [comment]
I believe that issue #584 is masking whatever the true problem is. The xml file should have a message in that case.

**Comment by agray** [comment]
Content of TestResult.xml is:

<?xml version="1.0" encoding="utf-8" standalone="no"?>
<test-run id="2" testcasecount="0" result="Passed" total="0" passed="0" failed="0" inconclusive="0" skipped="0" asserts="0" start-time="2015-04-09 11:38:59Z" end-time="2015-04-09 11:38:59Z" duration="0.094048">
  <environment nunit-version="3.0.5560.36865" clr-version="4.0.30319.34209" os-version="Microsoft Windows NT 6.3.9600.0" platform="Win32NT" cwd="C:\Jenkins\workspace\Saucery3NUnit" machine-name="ANDREWS-LAPTOP" user="andrewg" user-domain="andrews-laptop" culture="en-AU" uiculture="en-US" />
  <test-suite type="Assembly" id="1002" name="Saucery3Tester.dll" fullname="C:\Jenkins\workspace\Saucery3NUnit\NUnit\Version3\Saucery3Tester\bin\Release\Saucery3Tester.dll" runstate="Runnable" testcasecount="0" result="Passed" start-time="2015-04-09 11:38:59Z" end-time="2015-04-09 11:38:59Z" duration="0.038000" total="0" passed="0" failed="0" inconclusive="0" skipped="0" asserts="0">
    <properties>
      <property name="_PID" value="6944" />
      <property name="_APPDOMAIN" value="test-domain-" />
    </properties>
    <test-suite type="TestSuite" id="1003" name="Saucery3Tester" fullname="Saucery3Tester" runstate="Runnable" testcasecount="0" result="Passed" start-time="2015-04-09 11:38:59Z" end-time="2015-04-09 11:38:59Z" duration="0.021000" total="0" passed="0" failed="0" inconclusive="0" skipped="0" asserts="0">
      <test-suite type="SetUpFixture" id="1000" name="Tests" fullname="Saucery3Tester.Tests.GlobalSetup" classname="Saucery3Tester.Tests.GlobalSetup" runstate="Runnable" testcasecount="0" result="Passed" start-time="2015-04-09 11:38:59Z" end-time="2015-04-09 11:38:59Z" duration="0.005000" total="0" passed="0" failed="0" inconclusive="0" skipped="0" asserts="0" />
    </test-suite>
  </test-suite>
</test-run>

**Comment by agray** [comment]
GlobalSetup is:

    [OneTimeSetUp]
    public void RunBeforeAnyTests() {
    ...
    }

**Comment by CharliePoole** [comment]
It's interesting that the lowest level shown in the XML is your SetUpFixture, SauceryTestFixture.Tests.GlobalSetUp. No test fixture is detected. The class containing the three test cases doesn't seem to be found at all. What does the class declaration look like?

**Comment by CharliePoole** [comment]
So it appears that you are not a victim of #584 as I thought. Your XML result looks like everything was successfully loaded and passed - except of course that your test fixture and cases are missing!

**Comment by agray** [comment]
See updated comment above with namespace and class details, as requested

**Comment by CharliePoole** [comment]
I'm at a bit of a loss here, so I'll throw out some random suggestions:

1. Try the --explore option to see if the tests are never found at all or simply never executed.

2. Comment out the MethodImplAttributes - I'm wondering about both NUnit and .NET trying to control what methods can run simultaneously.

3. Use the --workers:1 or even --workers:0 option

If none of those experiments tell us anything, we're down to two ways to proceed:

* You build the debug version of NUnit from source and step through it.

* You give us a clean example that we can debug, eliminating any proprietary code.

Charlie

**Comment by agray** [comment]
1. I executed "C:\Program Files (x86)\NUnit.org\bin\nunit-console" --explore Saucery3Tester.dll from the command line and got the following output:

    NUnit Console Runner 3.0.5561
    Copyright (C) 2014 Charlie Poole

    Test Files
        Saucery3Tester.dll

    Runtime Environment
       OS Version: Microsoft Windows NT 6.3.9600.0
      CLR Version: 4.0.30319.34209

Did I execute --explore properly? What am I supposed to see?

**Comment by CharliePoole** [comment]
A list of tests. I guess it really isn't finding anything.

**Comment by agray** [comment]
I resolve an internal issue with my proprietary code and can current see the tests executing against saucelabs...so I reran the above command expecting to see a different result.  However, I did not.

**Comment by CharliePoole** [comment]
I'm on my phone so I can't try it out. Check the command-line help. It's a
slightly different code path so I thought it might show something.

Charlie

**Comment by agray** [comment]
I Tried a few combinations. Could only get output if I did /explore:nunit3

nunit3 file contains:

    <?xml version="1.0" encoding="utf-8" standalone="no"?>
    <test-run id="2" testcasecount="0">
      <test-suite type="Assembly" id="1002" name="Saucery3Tester.dll" fullname="C:\GitRepositories\saucery\NUnit\Version3\Saucery3Tester\bin\Release\Sauce
ry3Tester.dll" runstate="Runnable" testcasecount="0">
        <properties>
          <property name="_PID" value="2360" />
          <property name="_APPDOMAIN" value="test-domain-" />
        </properties>
        <test-suite type="TestSuite" id="1003" name="Saucery3Tester" fullname="Saucery3Tester" runstate="Runnable" testcasecount="0">
          <test-suite type="SetUpFixture" id="1000" name="Tests" fullname="Saucery3Tester.Tests.GlobalSetup" classname="Saucery3Tester.Tests.GlobalSetup"
runstate="Runnable" testcasecount="0" />
        </test-suite>
      </test-suite>
    </test-run>

However what I am seeing is the tests being run one at a time but GlobalSetup NOT being run.

**Comment by agray** [comment]
Current state of play:

- Tests being executed
- GlobalSetup class not being executed
- Test only executing one at a time (Parallelism not happening)

I have [assembly: Parallelizable(ParallelScope.Children)] and [assembly: LevelOfParallelism(16)] in AssemblyInfo.cs.  Is this the correct place to put them?

Is [this](https://github.com/nunit/dev/wiki/Framework-Parallel-Test-Execution) still current?

**Comment by CharliePoole** [comment]
--explore: I verified it works correctly. Without a file option it writes test cases one per line to the console, so no output if there are no tests. Probably could do with a message there.

test execution: are they invoked by NUnit as far as you can tell? Clearly nothing is being reported.

parallelism: I understood you only had one test fixture, so you can't expect any parallelism.

Attributes: The ones you have are fine. Assembly-level attributes can go anywhere, but in your AssembyInfo.cs is the most logical place.

Wiki specification: It's still the current specification :-) but it's neither documentation nor an as-built. Use the docs for what's actually in the build. For the most up to date docs, use the nunit wiki. For those pages not yet converted, use the web site. The parallel stuff is up to date in the wiki.

Have you had a chance to try my other suggestions?

**Comment by agray** [comment]
--explore is not producing any output.  However my tests do execute.  Given my tests are executing, the only problem remaining is that the tests are not running in parallel.  As per the link to the wiki in my last comment can I expect like the reporter did that at some point testcases will run in parallel in the same way that JUnit tests run in parallel under JUnit 4?

**Comment by CharliePoole** [comment]
I thought NUnit was not reporting the results of your tests either. Has that changed?

Parallel test cases are in the issue data base as milestone "Future"  - i.e. 3.2 or later. We decided that fixture parallelism was enough for most use cases and works most easily for users without changing their test code. OTOH test case parallelism will often break tests not designed for it.

**Comment by agray** [comment]
NUnit is emitting report xml file.  Jenkins NUnit plugin can't read it. I have raise [issue](https://issues.jenkins-ci.org/browse/JENKINS-27906).

That is very disappointing regarding Parallel test cases.  It should be implemented ASAP.

**Comment by CharliePoole** [comment]
Is the console report showing the correct results as well?

Regarding Jenkins, I think you need to use nunit2 output format in order to work with any of the CI builds. You can't reasonably expect them to parse the non-final form of the nunit3 output. That's why we included the nunit2 format as a feature.

**Comment by CharliePoole** [comment]
Re Parallel test cases: Too bad you weren't around arguing for it. :-)

**Comment by agray** [comment]
I can argue for it now.  Get its priority increased to the front of the queue.  I'm sure @rprouse is chomping at the bit to bring his MbUnit knowledge to bear.

**Comment by CharliePoole** [comment]
We can certainly always discuss. But our priority right now is to actually ship this  thing, not to increase the backlog!

Keep in mind that the only thing we can pull from MbUnit is the motivation to create a feature. Their code is not licensed in a way that we can use it, even if we wanted to.

More to the point... Is there anything left of the problem for which this bug report was written?

**Comment by agray** [comment]
Thanks Charlie, fair question.

While the tests are running. I am still interested to know why --explore produces no output from the commandline.

Based on your previous comment....you say --explore is producing output?

How?  What command syntax did you use?

Shall we close this and raise another issue?

**Comment by CharliePoole** [comment]
If you type
  nunit-console nunit.engine.tests.dll -explore
in the nunit bin directory, you'll get a list of the test cases in that
assembly.

What did you change to make your tests run? Maybe that will give us a hint.

If --explore remains broken, I think we can just redirect this to that
problem. It's a little used feature but if it doesn't work, there is likely
something going on that could impact execution at some time.

Charlie

BTW: This command-line stuff is documented fairly completely by
nunit-console --help

**Comment by agray** [comment]
For some reason the Console.WriteLine statements from GlobalSetup.cs are going to temporary-junit-reports\temp-junit.xml

Why would that be?  Under NUnit 2 the Console.WriteLine statements from this class are emitted to the console.

Why, under NUnit3 would it be written to a file?

**Comment by CharliePoole** [comment]
I have no idea. It sounds like Jenkins is trapping the output. All output
from a test is normally captured by NUnit and is added to the XML file. The
console displays that output upon receiving the test result. This is a
totally new approach from that of NUnit 2.x. I assume this output is _not_
showing up in the XML?

**Comment by agray** [comment]
The code works perfectly on Jenkins under NUnit 2.x.

Something somewhere is broken under NUnit 3.

I'll review the breaking changes page

**Comment by CharliePoole** [comment]
That page has things we broke on purpose. :-)

Are you sure your tests are running perfectly under nunit-console, without Jenkins? You'll want to separate the impact of NUnit's interaction with your product from it's interaction with Jenkins.

**Comment by jmbeach** [comment]
God, took me forever to realize I had visual studio in Release mode....
